Storage: Add NVMe and SDC storage connectors

MusicDin opened the pull request:
This PR contains cherry-picks from #14700 to include only connectors and PowerFlex changes (without Pure Storage bits)

The aim of the connector is to handle connections with the target storage. Currently, connectors support NVMe, and SDC.

Once PowerFlex will use `Connect`/`Diconnect`, the following connector functions can be removed:
- `ConnectAll`
- `DisconnectAll`
- `SessionID`

Closes https://github.com/canonical/lxd/pull/14708

MusicDin commented:
As previously discussed with @roosterfish, I have also replaced nvme "disconnect-all" and "connect-all" by connecting to just a specific target. This prevents disconnecting all nvme volumes, which would affect other storage drivers as well. However, needs to be checked that NQN is correctly passed.

roosterfish commented on lxd/storage/connectors/connector_nvme.go:
```diff
@@ -0,0 +1,171 @@
+}
+
+// ConnectAll establishes a connection with all targets available on the given address.
+func (c *connectorNVMe) ConnectAll(ctx context.Context, targetAddr string) error {
```
I am about to suggest that we can drop the `ConnectAll` and `DisconnectAll` methods and replace the ones in PowerFlex with the respective `Connect` and `Disconnect`. 
In case of PowerFlex the SDT (NVMe targets) can be queried easily so that shouldn't be too much of an issue.

But I think this can land right after this PR so we don't have to mix up things.

roosterfish replied:
Maybe `ConnectAll` can be kept for convenience, but really PowerFlex shouldn't use `DisconnectAll` anymore.

MusicDin replied:
Sure, connect-all should not interfere. Will fix

roosterfish commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -931,19 +914,26 @@ func (d *powerflex) mapVolume(vol Volume) (revert.Hook, error) {
+	targetQN := domain.SystemID
+	targetAddr := d.config["powerflex.sdt"]
+
+	err = d.connector().Connect(d.state.ShutdownCtx, targetAddr, targetQN)
```
`ConnectAll`?

roosterfish commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -931,19 +914,26 @@ func (d *powerflex) mapVolume(vol Volume) (revert.Hook, error) {
+		return nil, err
 	}
 
+	reverter.Add(func() { _ = d.connector().Disconnect(targetQN) })
```
`DisconnectAll`?

MusicDin replied:
I would avoid DisconnectAll, as previously discussed, this would disconnect all NVMe volumes from the LXD host, which is not desired.

roosterfish commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -810,13 +784,17 @@ func (d *powerflex) getVolumeType(vol Volume) powerFlexVolumeType {
 	var hostID string
-	nqn := d.getHostNQN()
+
+	targetNQN, err := d.connector().QualifiedName()
```
`hostNQN` as target would be PowerFlex?

MusicDin replied:
Should be hostNQN. This is mistake

roosterfish commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -847,8 +825,13 @@ func (d *powerflex) createNVMeHost() (string, revert.Hook, error) {
-	nqn := d.getHostNQN()
-	host, err := client.getNVMeHostByNQN(nqn)
+
+	targetNQN, err := d.connector().QualifiedName()
```
`hostNQN`? Same as above.

roosterfish commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -1114,28 +1041,35 @@ func (d *powerflex) unmapVolume(vol Volume) error {
 			// Disconnect from the NVMe subsystem.
 			// Do this first before removing the host from PowerFlex.
-			err := d.disconnectNVMeSubsys()
+			err = d.connector().Disconnect(targetQN)
```
`DisconnectAll`?

MusicDin replied:
I would avoid DisconnectAll, as previously discussed, this would disconnect all NVMe volumes from the LXD host, which is not desired.

roosterfish reviewed (changes requested):
Thanks for this, looking great already. 

I'll grab your branch and make some tests on the PowerFlex cluster for validation. I think the only change that has to be performed upfront is using the actual (Connect|Disconnect)All funcs to resemble the original behavior.
When we have merged the PR I can take care of switching to the Connect/Disconnect instead.

tomponline commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -885,7 +885,7 @@ func (d *powerflex) mapVolume(vol Volume) (revert.Hook, error) {
 	switch d.config["powerflex.mode"] {
 	case powerFlexModeNVMe:
-		unlock, err := locking.Lock(d.state.ShutdownCtx, "nvme")
+		unlock, err := locking.Lock(d.state.ShutdownCtx, "storage_powerflex_nvme")
```
we've got `imageOperationLock(fingerprint string)` which we could have something similar, e.g. `storageConnectLock(connectorName string, driverName string)`

MusicDin commented:
@roosterfish Thanks for the comments. I've addressed all of them, and reverted fixes for `ConnectAll -> Connect`, and `DisconnectAll -> Disconnect` to retain the previous implementation.

roosterfish commented:
> I've addressed all of them, and reverted fixes

Thanks, I'll run it through some tests on the PowerFlex env and report back here.

MusicDin commented on lxd/storage/connectors/utils.go:
```diff
@@ -0,0 +1,126 @@
+		}
+
+		// Skip volumes that do not contain the desired name suffix.
+		if !strings.HasSuffix(diskPath, diskNameSuffix) {
```
Good point, accepting an arbitrary function also seems versatile solution for any potential upcoming storage drivers

MusicDin replied:
I've added the `devicePathFilter` which is a function that accepts device path, and returns bool indicating whether the path matches the required criteria.

roosterfish commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -946,19 +913,19 @@ func (d *powerflex) mapVolume(vol Volume) (revert.Hook, error) {
+	// In case of NVMe/TCP, we have to connect after the first mapping was established,
+	// as PowerFlex does not offer any discovery log entries until a volume gets mapped
+	// to the host.
+	err = d.connector().ConnectAll(d.state.ShutdownCtx, targetAddr)
```
Ok, I made progress in the test suite running those changes on PowerFlex. 
Whilst doing so I found a bug as the environment is quite slow :) But wanted to check if it's related to the changes in this PR, but it's something different.

Anyway the `lxd-ci/tests/storage-vm` is now running until vol resizing which fails. I tried to remember what the error was about as it seems familiar, and if I am not mistaken it has something to do with this line here.

Prior to this change the `connectNVMeSubsys()` func was returning a cleanup hook which only contained the disconnect instructions in case the call to `connectNVMeSubsys` actually caused a connection (as it is idempotent and the connection might already be established).
If we are already connected to the subsystem, the returned cleanup hook didn't contain the disconnect statement. Now it does in any case.
IIRC that was one of the problems I found and fixed in https://github.com/canonical/lxd/pull/13550.
And I assume this is now causing the resizing to fail as LXD rips out the connection whilst trying to resize the vol.
It calls the `SetVolumeQuota` func which tries to `mapVolume`. Before this PR the returned cleanup hook didn't contain the disconnect but now it does. So if I get the timing right the resize succeeds, but then the subsystem gets disconnected but LXD tries to bring up the VM again which fails. 

If you like let's have a short meeting to discuss this. Hope I was able to explain it well enough.

roosterfish replied:
Maybe that is also something to check for the pure driver? Simply start a VM, resize its root vol and the do a `lxc restart <vm>`. This should cause the error if pure is affected too.

roosterfish commented on lxd/storage/connectors/connector.go:
```diff
@@ -0,0 +1,76 @@
+		}
+
+	case TypeSDC:
+		return &connectorNVMe{
```
Should be `connectorSDC` :)

roosterfish commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -946,17 +915,36 @@ func (d *powerflex) mapVolume(vol Volume) (revert.Hook, error) {
 		reverter.Add(func() { _ = client.deleteHostVolumeMapping(hostID, volumeID) })
```
Here is the summary of changes:

```go
    // Add this new outerRevert which handles reverts we can safely pass to the caller.
	outerRevert := revert.New()
	defer outerRevert.Fail()

	if !mapped {
		err = client.createHostVolumeMapping(hostID, volumeID)
		if err != nil {
			return nil, err
		}

        // Run umapVolume instead as it contains the logic required to know if the caller can still delete the NVMe host and close the NVMe/TCP connection. Furthermore it aquires the lock to safely perform those actions.
		outerRevert.Add(func() { _ = d.unmapVolume(vol) })
	}

    ...

    // Get the cleanup from the outerRevert
    cleanup := outerRevert.Clone().Fail

    // Set success for the inner revert as deletion of the host is something we don't want the caller to run as it might already affect other volumes as we have given up the lock between returning the cleanup and actually calling it.
	reverter.Success()
	outerRevert.Success()
	return cleanup, nil
```

roosterfish commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -946,17 +915,36 @@ func (d *powerflex) mapVolume(vol Volume) (revert.Hook, error) {
+		return nil, err
+	}
+
+	if sessionID == "" {
```
As we add `unmapVolume` above in the reverter we don't need to add the disconnect to the reverter as this is handled within `unmapVolume` after properly acquiring the lock.

roosterfish reviewed (changes requested):
Thanks for the short call, here a summary of what we have discussed:

roosterfish reviewed (approved):
Thanks, just passed lxd-ci using nvme mode. LGTM!

tomponline commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -946,22 +915,27 @@ func (d *powerflex) mapVolume(vol Volume) (revert.Hook, error) {
+	// because it ensures the lock is acquired and accounts for
+	// an existing session before unmapping a volume.
+	outerReverter := revert.New()
+	outerReverter.Add(func() { _ = d.unmapVolume(vol) })
```
My understanding from @MusicDin is that `unapVolume` already handles that logic?

roosterfish replied:
I have just discussed this with @MusicDin. 
Within `unmapVolume` you have already lost the information whether or not you or somebody else has mapped the volume. 
All you can see is if the volume is is mapped to the host or not. 
Whereas in `mapVolume` you still have the knowledge if you actually performed the mapping because you obtain the lock and then check if the volume is currently mapped to the host.

MusicDin replied:
In some cases, such as online resizing of the volume, we attempt to map the volume even if it is already mapped.

If an error occurs in the above case, assuming `mapVolume` always returns `unmapVolume` in the reverter, it would cause the host to be disassociated from the volume (on remote storage). As a result the volume would disappear from the host even if the connection with the remote storage is still established. (disassociating the volume can be though of as removing permission from the client to access the volume, not breaking the connection)

However, `unmapVolume` will only prevents disconnection if there are still volumes associated with the host (after the desired one is disassociated).

With that being said, @roosterfish is correct, and we should only ever revert the mapping if the volume was just mapped.

tomponline replied:
@MusicDin @roosterfish so whats the outcome?

tomponline replied:
and in other cases we return an empty reverter that does nothing right?

roosterfish commented:
NVMe Test passed with latest changes!

tomponline commented on lxd/storage/connectors/utils.go:
```diff
@@ -0,0 +1,128 @@
+		}
+
+		// Check if context is cancelled.
+		if ctx.Err() != nil {
```
small nit that we are calling ctx.Err() twice unnecessarily when it is non-nil.

tomponline commented on lxd/storage/connectors/utils.go:
```diff
@@ -126,3 +130,120 @@ func WaitDiskDeviceGone(ctx context.Context, diskPath string) bool {
+		// connections are closed. Otherwise, keep open connections
+		// intact for other volumes using the session.
+		if session == nil {
+			revert.Add(func() { _ = c.Disconnect(targetQN) })
```
@MusicDin this looks wrong, you're adding stuff to the reverter that only is added after the revert.Fail function is run.

MusicDin replied:
Oops

tomponline commented on lxd/storage/connectors/utils.go:
```diff
@@ -126,3 +130,120 @@ func WaitDiskDeviceGone(ctx context.Context, diskPath string) bool {
+	// exit if all connections fail.
+	doneConns := 0
+	for {
+		success := <-doneChan
```
When the go routine that calls `wg.Wait()` ends, it closes `doneChan` which will cause all reads from that channel to return `false` (the default bool value), so I think we can just `range` over this channel, and return the first true value or wait until the for range loop ends (due to the channel being closed), and then there's no need for the `doneConns` logic.

tomponline replied:
So like this (also suggest rename to resChan for clarity, as `doneChan` usually indicates its closed when done only):

```go
for success := range resChan {
     if success {
        return nil // Successfully connected to one endpoint
     }
}

return nil, fmt.Errorf("Failed to connect to any address on target %q", targetQN)
```

tomponline replied:
Further more if you made `resChan` type `error` you could do this, and return the last error to the caller.

```go
var err error
for err = range resChan {
     if err == nil {
        return nil // Successfully connected to one endpoint
     }
}

return fmt.Errorf("Failed to connect to any address on target %q: %w", targetQN, err)
```

tomponline commented on lxd/storage/connectors/connector.go:
```diff
@@ -37,3 +40,24 @@ type Connector interface {
+			common: common,
+		}
+
+	default:
```
why wouldn't we expect the caller to do this, or for this function to return an error if an invalid connectorType string is provided? Seems like its going to go wrong sooner or later anyway, and sooner is preferable right?

MusicDin replied:
The main idea behind that was to avoid returning error from `driver.Connector()`. The mode is always checked before initializing the connector, and to avoid checking that error is not nil (or ignoring it) across utils function, I have instead returned the common connector in such case. The common connector errors out for almost all connector functions anyway and returns type "unknown".

In short to be able to do:
```go
if d.connector().Type() == connectors.NVME {
   ...
}

or 

switch d.connector().Type() {
 ...
}
```

Instead of:
```go
con, err := d.connector().Type()
if err != nil {
  return err
}

if con.Type() == connectors.NVME {
   ...
}

or 

con, _ := d.connector().Type()
switch con.Type() {
 ...
}
```

If the error is not ignored, we would always need to either  "nil" error

It allows then to do:
```
err := NewConnector(conntectors.TypeNVME).LoadModules()
```

So if it is acceptable, we can return error from `Connector()` and still ignore it in `driver.connector()` wrapper?

tomponline replied:
My view is that `NewConnector` should return nil with an error if an invalid type is provided, like the driver loaders do. https://github.com/canonical/lxd/blob/main/lxd/instance/drivers/load.go#L65

tomponline commented on lxd/storage/connectors/connector_nvme.go:
```diff
@@ -0,0 +1,261 @@
+			return fmt.Errorf("Failed to connect to target %q on %q via NVMe: %w", targetQN, targetAddr, err)
+		}
+
+		if stderr != "" {
```
could do with a comment here explaining in what scenarios we'd expect to get a zero (success) exit code with non-empty stderr?

MusicDin replied:
To be sincere, I am not sure if there is any such case. Since I never ran into any issues with that, I assumed that it will either be skipped immediately or show some other underlying error.

But now that you have pointed it out, I think this can make more harm than good in case some warnings are printed to stderr and the command exits successfully. This would result in count as the *connection failed* although if exit status is 0, I quite sure the connection succeeded.

@roosterfish Do you have any specific use case for that, otherwise I'll happily remove this.

roosterfish replied:
I don't remember why this was useful. But I chose `RunCommandSplit` for a reason to explicitly get `stderr`. 
Maybe it was for testing purposes so I would say let's get rid of it until we find a use case we can then document.

tomponline commented on lxd/storage/connectors/connector_nvme.go:
```diff
@@ -0,0 +1,261 @@
+		return fmt.Errorf("Failed to connect to any target on %q via NVMe: %w", targetAddr, err)
+	}
+
+	if stderr != "" {
```
could do with a comment here explaining in what scenarios we'd expect to get a zero (success) exit code with non-empty stderr?

tomponline commented on lxd/storage/connectors/utils.go:
```diff
@@ -126,3 +130,120 @@ func WaitDiskDeviceGone(ctx context.Context, diskPath string) bool {
+	doneChan := make(chan bool, len(targetAddrs))
+
+	// Connect to all target addresses.
+	for _, addr := range targetAddrs {
```
I think I'd be tempted to move this into the same go routine as the cleanup routine below so that its keeps these parts locked together and more readable, e.g.

```go
go func() {
	// Connect to all target addresses.
	for _, addr := range targetAddrs {
          go func(addr string) {
           ....
          })(addr)
    }

        wg.Wait()
		close(doneChan)
		cancel()
		unlock()
}()
```

tomponline commented on lxd/storage/drivers/utils.go:
```diff
@@ -911,3 +912,15 @@ func ResolveServerName(serverName string) (string, error) {
+	l.Debug("Acquiring lock for storage connector")
+	defer l.Debug("Lock acquired for storage connector")
+
+	return locking.Lock(context.TODO(), fmt.Sprintf("StorageConnector_%s_%s", connectorName, driverName))
```
Can the logging and lock name be updated to reflect the new function name please.

tomponline commented on lxd/storage/drivers/driver_powerflex_utils.go:
```diff
@@ -885,7 +885,7 @@ func (d *powerflex) mapVolume(vol Volume) (revert.Hook, error) {
 	switch d.config["powerflex.mode"] {
 	case powerFlexModeNVMe:
-		unlock, err := locking.Lock(d.state.ShutdownCtx, "nvme")
+		unlock, err := locking.Lock(d.state.ShutdownCtx, "storage_powerflex_nvme")
```
would be good to have constants for these so we can see where they are being referenced?

MusicDin replied:
This commit can be dropped. It is a leftover, as this is fixed in one of the following commits using `remoteVolumeMapLock`

tomponline reviewed (commented):
Nearly there, few small changes to make. Ta

MusicDin commented on lxd/storage/connectors/utils.go:
```diff
@@ -127,3 +131,113 @@ func WaitDiskDeviceGone(ctx context.Context, diskPath string) bool {
+		cancel()
+
+		// Wait until all connection attempts have finished.
+		wg.Wait()
```
Hmm, this one is tricky. It *could* happen. But how long should we keep it?

MusicDin replied:
Having a counter would be a viable solution. But when do we release the initial lock/reduce the counter

tomponline replied:
@MusicDin @roosterfish and I discussed it and we'll merge this as-is, and then @roosterfish is going to see if he can address this with his other up coming PowerFlex fixes.

That way we can unblock the Pure driver.

roosterfish commented on lxd/storage/connectors/utils.go:
```diff
@@ -0,0 +1,243 @@
+		// connections are closed. Otherwise, keep open connections
+		// intact for other volumes using the session.
+		if session == nil {
+			_ = c.Disconnect(context.Background(), targetQN)
```
Instead should this be the `ctx` passed inside `connect()`?

MusicDin replied:
I don't think so because the caller may use context with shorter timeout for connect, so we would never run disconnect if the timeout is reached.

tomponline replied:
no, i dont think so because it'll be guaranteed cancelled as its cancelled above on line 217.

MusicDin replied:
I would keep this question open until we figure out the previous issue, which may solve this one as well - we may end up not returning the reverter at all (or at least not disconnecting in the reverter)

roosterfish reviewed (approved):
Ran another bunch of tests on NVMe/SDC with VMs and containers. All good and cleaned up afterwards. LGTM!

tomponline reviewed (approved):
LGTM thanks!
